<h1>Data Cleaning / Munging</h1>

<p>In these notes we look at creating various plots based on a set of data.  We have a datafile called degreeClassification.csv (available on blackboard) which consists of the following columns:</p>
<ul>
<li>Academic Year:  The year that the degree was awarded.</li>
<li>Mode of study: Full-time, part-time or both.</li>
<li>Sex: Male, Female, Other or all.</li>
<li>Country of HE provider: England, Northern Ireland, Scotland, Wales or all.</li>
<li>Classification of First Degree: First, Upper Second, Lower Second or Third/Pass.</li>
<li>Number: The number of students achieving that degree.</li>
<li>Percentage:  The percentage of students achieving that degree.</li>
</ul>   

<p>We start the exercises by getting the data ready to be plotted</p>

<h2>Import PANDAs</h2>

In [1]:
import pandas as pd;
import re;

<h2>Read the data into a PANDAs DataFrame</h2>
<p>Read the 'degreeClassification.csv' in to a data frame called df, use the head method to output the first 20 rows of the datafile so that the structure of the dataframe can be explored.</p>

In [2]:
df = pd.read_csv("degreeClassification.csv")
df.head(20)

Title  \
0                   Location   
1             Academic years   
2                Data source   
3           Data source link   
4   Data file canonical link   
5                    Licence   
6                  Code page   
7                 Disclaimer   
8              2013/14 total   
9              2014/15 total   
10             2015/16 total   
11             2016/17 total   
12             2017/18 total   
13              Last updated   
14                       NaN   
15                       NaN   
16             Academic Year   
17                   2013/14   
18                   2013/14   
19                   2013/14   

   Percentage of first degree qualifiers obtaining each classification by sex, mode of study and country of HE provider  \
0                                                  UK                                                                     
1                                  2013/14 to 2017/18                                                                     
2                                                HESA                                                                     
3   https://www.hesa.ac.uk/data-and-analysis/sb252...                                                                     
4   https://www.hesa.ac.uk/data-and-analysis/sb252...                                                                     
5   Creative Commons Attribution 4.0 International...                                                                     
6                                      Unicode UTF-8                                                                      
7   Please note that this data includes rounded to...                                                                     
8                                              395605                                                                     
9                                              370905                                                                     
10                                             376330                                                                     
11                                             391875                                                                     
12                                             396940                                                                     
13                                             Jan-19                                                                     
14                                                NaN                                                                     
15                                                NaN                                                                     
16                                      Mode of study                                                                     
17                                                All                                                                     
18                                                All                                                                     
19                                                All                                                                     

   Unnamed: 2              Unnamed: 3                      Unnamed: 4  \
0         NaN                     NaN                             NaN   
1         NaN                     NaN                             NaN   
2         NaN                     NaN                             NaN   
3         NaN                     NaN                             NaN   
4         NaN                     NaN                             NaN   
5         NaN                     NaN                             NaN   
6         NaN                     NaN                             NaN   
7         NaN                     NaN                             NaN   
8         NaN                     NaN                             NaN   
9         NaN                     Na

<h2>Skipping Summary Rows</h2>
<p>Read the datafile and skip the rows that contain summary information.  Use the head method and output the first 20 rows.  The column names should be:
<ul>    
<li>Academic Year</li>
<li>Mode of study</li>
<li>Sex</li>
<li>Country of HE provider</li>
<li>Classification of First Degree</li>
<li>Number</li>
<li>Percentage</li>
</ul>    
</p>

In [3]:
df = pd.read_csv("degreeClassification.csv", skiprows=17)
df.head(20)

Academic Year Mode of study  Sex Country of HE provider  \
0        2013/14           All  All                    All   
1        2013/14           All  All                    All   
2        2013/14           All  All                    All   
3        2013/14           All  All                    All   
4        2013/14           All  All                    All   
5        2013/14           All  All                England   
6        2013/14           All  All                England   
7        2013/14           All  All                England   
8        2013/14           All  All                England   
9        2013/14           All  All                England   
10       2013/14           All  All       Northern Ireland   
11       2013/14           All  All       Northern Ireland   
12       2013/14           All  All       Northern Ireland   
13       2013/14           All  All       Northern Ireland   
14       2013/14           All  All       Northern Ireland   
15       2013/14           All  All               Scotland   
16       2013/14           All  All               Scotland   
17       2013/14           All  All               Scotland   
18       2013/14           All  All               Scotland   
19       2013/14           All  All               Scotland   

   Classification of First Degree  Number Percentage  
0                           First   79410        20%  
1                    Lower second   95965        24%  
2                      Third/Pass   21905         6%  
3                Total classified  395605        NaN  
4                    Upper second  198320        50%  
5                           First   68590        20%  
6                    Lower second   81665        24%  
7                      Third/Pass   19385         6%  
8                Total classified  339170        NaN  
9                    Upper second  169530        50%  
10                          First    1590        19%  
11                   Lower second    2215        26%  
12                     Third/Pass     285         3%  
13               Total classified    8560        NaN  
14                   Upper second    4470        52%  
15                          First    5125        21%  
16                   Lower second    5505        22%  
17                     Third/Pass     875         4%  
18               Total classified   24880        NaN  
19                   Upper second   13380        54%

<h2>Unique values in a Column</h2>
<p>Output the unique values for the column 'Mode of study' to give a general idea of the categorical data for this column</p>

In [4]:
def print_unique_modes():

    modes_of_study = df["Mode of study"].unique()
    for mode in modes_of_study:
        print(mode)
        
print_unique_modes()

All
Full-time
Part-time


<h2>Boolean Masks</h2>

<p>Create a boolean mask and then filter the data based on the categorical data value 'All' for 'Mode of Study'.  We do this in stages to help understand how the more complex aggregated instruction is built.</p>
<p>The overall goal is to apply a mask so that all the rows that contain values part-time or full-time are dropped from the dataframe.  We do this so that we don’t double count, i.e. A count in 'All' will also also be a count in either 'Part-time' or 'Full-time'.</p>

<ol>
<li>Create a mask to output True/False for each row, True if 'Mode of study' is equal to 'All' and False otherwise</li>

In [5]:
modeMask = df["Mode of study"] == "Full-time"


<ol start='2'>
<li>Assign the Boolean mask to a variable named mask</li>

In [6]:
mask = modeMask

<ol start='3'>
<li>Use the variable named mask to update the dataframe named df so that the dataframe is filtered based on the boolean mask</li>
<li>Output the unique values for the column 'Mode of study' to ensure that 'Full-time' and 'Part-time' have been filtered out (don't exist in the dataframe)</li>

In [7]:
df = df[mask]
print_unique_modes()

Full-time


<ol start='5'>
<li>Output the dataframe</li>

In [8]:
df

Academic Year Mode of study    Sex Country of HE provider  \
95         2013/14     Full-time    All                    All   
96         2013/14     Full-time    All                    All   
97         2013/14     Full-time    All                    All   
98         2013/14     Full-time    All                    All   
99         2013/14     Full-time    All                    All   
...            ...           ...    ...                    ...   
1290       2017/18     Full-time  Other                  Wales   
1291       2017/18     Full-time  Other                  Wales   
1292       2017/18     Full-time  Other                  Wales   
1293       2017/18     Full-time  Other                  Wales   
1294       2017/18     Full-time  Other                  Wales   

     Classification of First Degree  Number Percentage  
95                            First   73810        20%  
96                     Lower second   85760        24%  
97                       Third/Pass   16670         5%  
98                 Total classified  362370        NaN  
99                     Upper second  186125        51%  
...                             ...     ...        ...  
1290                          First      10         ..  
1291                   Lower second       0         ..  
1292                     Third/Pass       0         ..  
1293               Total classified      20        NaN  
1294                   Upper second      10         ..  

[475 rows x 7 columns]

<h2>Boolean Masks Continued</h2>

<p>Create a boolean mask and then filter the data based on the categorical data 'All' for 'Sex'</p>
<p>Then create a boolean mask and then filter the data based on the categorical data 'All' for 'Country of HE provider'</p>
<p>Output the dataframe to check the filtering was successful, you should have reduced the data from 1380 rows to 25 rows.</p>

In [9]:
df = df[df["Sex"] == "All"]
df = df[df["Country of HE provider"] == "All"]
df.shape

(25, 7)

<h2>More Boolean Masks</h2>

<p>We are not interested in rows that show the totals of classification for each year, filter these out.</p>
<p>Output the dataframe to check the filtering was successful, you should have reduced the data from 25 rows to 20 rows.</p>

In [10]:
df = df[df["Classification of First Degree"] != "Total classified"]
df.shape

(20, 7)

<h2>Dropping Columns</h2>
<p>Since columns 'Mode of study', 'Sex', and 'Country of HE provider' don\t provide us with any meaninful information (the all contain the data 'All'), drop these columns from the dataframe.</p>
<p>Output the dataframe to check the columns have been dropped successfully.</p>

In [11]:
df.drop(["Sex", "Mode of study", "Country of HE provider"], axis=1, inplace=True)
df

Academic Year Classification of First Degree  Number Percentage
95         2013/14                          First   73810        20%
96         2013/14                   Lower second   85760        24%
97         2013/14                     Third/Pass   16670         5%
99         2013/14                   Upper second  186125        51%
370        2014/15                          First   76030        22%
371        2014/15                   Lower second   75350        22%
372        2014/15                     Third/Pass   15080         4%
374        2014/15                   Upper second  171705        51%
645        2015/16                          First   83720        24%
646        2015/16                   Lower second   73180        21%
647        2015/16                     Third/Pass   14335         4%
649        2015/16                   Upper second  176215        51%
920        2016/17                          First   95525        26%
921        2016/17                   Lower second   71290        20%
922        2016/17                     Third/Pass   14225         4%
924        2016/17                   Upper second  182305        50%
1200       2017/18                          First  105500        28%
1201       2017/18                   Lower second   69105        19%
1202       2017/18                     Third/Pass   13435         4%
1204       2017/18                   Upper second  184150        49%

<h2>Checking the Data Type of a Column</h2>
<p>Check the type of data that is stored for one of the values in the percentage column.</p>

In [12]:
type(df.loc[95]["Percentage"])

str

<h2>Cleaning and converting to an string data type to a number data type</h2>
<p>The percentage column is a string data type because of the % character.  Remove the % character and convert the values for this column to a numeric data type</p>

In [13]:
df["Percentage"] = df["Percentage"].str.replace("%", "").astype(int)
type(df.loc[95]["Percentage"])

numpy.int32

<h2>New Dataframes</h2>
<p>Often it is useful to create seperate smaller dataframes, create four new dataframes named df_first, df_upp_sec, df_low_sec and df_third, to seperate the different degree classifications.</p>

Example:
df_first would output:
<pre>
      Academic Year   Classification of First Degree	Number	 Percentage
0	    2013/14	       First	                    79410	  20%
275	  2014/15	       First	                    81640	  22%
550	  2015/16	       First	                    88890	  24%
825	  2016/17	       First	                    100945	 26%
1105	 2017/18	       First	                    110475	 28%
</pre>

In [14]:
df_first = df[df["Classification of First Degree"] == "First"]
df_upp_sec = df[df["Classification of First Degree"] == "Upper second"]
df_low_sec = df[df["Classification of First Degree"] == "Lower second"]
df_third = df[df["Classification of First Degree"] == "Third/Pass"]

<h2>Save the Dataframes to CSV files</h2>
<p>Save the dataframes df, df_first, df_upp_sec, df_low_sec and df_third to csv files with filesnames that match the dataframe name.

In [15]:
df_first.to_csv("df_first.csv")
df_upp_sec.to_csv("df_upp_sec.csv")
df_low_sec.to_csv("df_low_sec.csv")
df_third.to_csv("df_third.csv")
df.to_csv("df.csv")